stdlib/public/core/Integers.swift.gyb

```diff
@@ -2483,6 +2483,172 @@ ${assignmentOperatorComment(x.operator, False)}
   % end
 }
 
+% for Range in ['Range', 'ClosedRange']:
+%   exampleRange = '1..<100' if Range == 'Range' else '1...100'
+
+extension ${Range}
+  where Bound: FixedWidthInteger, Bound.Stride : SignedInteger,
+  Bound.Magnitude: UnsignedInteger
+{
+      
+  /// Returns a random element of the range, using the given generator as
+  /// a source for randomness.
+  ///
+  /// You can use this method to select a random element of a range when you
+  /// are using a custom random number generator. If you're generating a random
+  /// number, in most cases, you should prefer using the `random(in:using:)`
+  /// static method of the desired numeric type. That static method is available
+  /// for both integer and floating point types, and returns a non-optional
+  /// value.
+  ///
+  /// - Parameter generator: The random number generator to use when choosing
+  ///   a random element.
+  /// - Returns: A random element of the range.
+%   if 'Closed' not in Range:
+  ///   If the range is empty, the method returns `nil`.
+%   else:
+  ///   This method never returns `nil`.
+%   end
+  @inlinable
+  public func randomElement<T: RandomNumberGenerator>(
+    using generator: inout T
+  ) -> Element? {
```

Comment:
I think the following might be possible, to avoid the GYB code.

```swift
extension ${Range} where
  Bound: FixedWidthInteger,
  Bound.Stride: SignedInteger,
  Bound.Magnitude: UnsignedInteger
{

  internal var _magnitude: Bound.Magnitude? {
    guard !(contains(.min) && contains(.max)) else {
      return nil
    }
    let magnitude: Bound.Magnitude
    switch (lowerBound.signum(), upperBound.signum()) {
    case (-1, -1):
      magnitude = lowerBound.magnitude - upperBound.magnitude
    case (1, 1):
      magnitude = upperBound.magnitude - lowerBound.magnitude
    default:
      magnitude = lowerBound.magnitude + upperBound.magnitude
    }
    return contains(upperBound) ? magnitude + 1 : magnitude
  }

  public func randomElement<T: RandomNumberGenerator>(
    using generator: inout T
  ) -> Element? {
    guard !isEmpty else {
      return nil
    }
    let randomDistance: Bound.Magnitude
    if let maxDistance = _magnitude {
      randomDistance = generator.next(upperBound: maxDistance)
    } else {
      randomDistance = generator.next()
    }
    return lowerBound &+ Bound(truncatingIfNeeded: randomDistance)
  }
}
```

The overflow addition is a bit obscure, but it seems to work.

Is there a `Bound.bitWidth == Bound.Magnitude.bitWidth` guarantee?

Reply:
I don't think we need to treat avoiding GYB in a GYB'd file as a goal, particularly if it ends up adding runtime branches. If it makes sense to extract some of this to a common routine, that's great, but GYB should generate the code you would write if you were writing out each version specifically. (Here and below.)